File: python/hqc/rm.py
"""Reed-Muller RM(1,7) encode and decode.

RM(1,7) encodes 8 bits (1 byte) into 128 bits.
The generator matrix has rows corresponding to:
- Row 0: all-ones vector (length 128)
- Rows 1..7: the i-th bit of each position index

Decoding uses the Walsh-Hadamard transform to find the closest codeword.
With multiplicity, each byte is encoded into n2 = 128 * multiplicity bits.
"""


def rm_encode_single(msg_byte: int) -> int:
    """Encode a single byte using RM(1,7) into a 128-bit integer.

    msg_byte has 8 bits: b0 is the constant term, b1..b7 are the linear terms.
    The codeword c[j] for j in [0,128) is:
        c[j] = b0 XOR (b1 & j_0) XOR (b2 & j_1) XOR ... XOR (b7 & j_6)
    where j_i is bit i of j.
    """
    result = 0
    b0 = (msg_byte >> 0) & 1
    for j in range(128):
        bit = b0
        for i in range(7):
            if (msg_byte >> (i + 1)) & 1:
                bit ^= (j >> i) & 1
        if bit:
            result |= (1 << j)
    return result


def rm_encode(msg_byte: int, multiplicity: int) -> int:
    """Encode a byte using RM(1,7) with repetition (multiplicity).

    Returns n2 = 128 * multiplicity bits as an integer.
    The 128-bit codeword is repeated 'multiplicity' times.
    """
    codeword = rm_encode_single(msg_byte)
    result = 0
    for i in range(multiplicity):
        result |= (codeword << (128 * i))
    return result


def _walsh_hadamard_transform(v: list) -> list:
    """In-place Walsh-Hadamard transform on a list of length 128 (= 2^7).

    Transforms real-valued vector v such that v[i] = sum_j (-1)^(popcount(i&j)) * v[j].
    """
    n = len(v)
    h = 1
    while h < n:
        for i in range(0, n, h * 2):
            for j in range(i, i + h):
                x = v[j]
                y = v[j + h]
                v[j] = x + y
                v[j + h] = x - y
        h *= 2
    return v


def rm_decode(received: int, n2: int, multiplicity: int) -> int:
    """Decode a received n2-bit word to recover the original byte.

    Uses Walsh-Hadamard transform on the sum of all multiplicity copies.
    received is an integer with n2 bits.
    """
    # Sum the multiplicity copies (convert bits to +1/-1 and accumulate)
    sums = [0] * 128
    for m in range(multiplicity):
        block = (received >> (128 * m)) & ((1 << 128) - 1)
        for j in range(128):
            bit = (block >> j) & 1
            # Map 0 -> +1, 1 -> -1
            sums[j] += 1 - 2 * bit

    # Apply Walsh-Hadamard transform
    wht = _walsh_hadamard_transform(sums)

    # Find the index with maximum absolute value
    best_val = 0
    best_idx = 0
    for i in range(128):
        if abs(wht[i]) > abs(best_val):
            best_val = wht[i]
            best_idx = i

    # Recover the message byte
    # best_idx gives bits 1..7 of the message
    # The sign of best_val gives bit 0: positive means b0=0, negative means b0=1
    msg_byte = 0
    if best_val < 0:
        msg_byte = 1  # b0 = 1

    # Bits 1..7 come from the index
    for i in range(7):
        if (best_idx >> i) & 1:
            msg_byte |= (1 << (i + 1))

    return msg_byte

File: go/internal/hqc/rm.go
package hqc

// Reed-Muller code RM(1, 7) for HQC.
//
// The first-order Reed-Muller code RM(1, 7) encodes 8 bits (1 byte)
// into 128 bits (16 bytes). The encoding uses the generator matrix
// of RM(1, 7), which consists of the all-ones word and the 7 rows
// of the Walsh-Hadamard basis vectors.
//
// For HQC, the RM codeword is then duplicated (Multiplicity times)
// to form an n2-bit codeword for additional error correction.

// rmBaseLen is the base Reed-Muller codeword length = 2^7 = 128 bits.
const rmBaseLen = 128

// rmEncodeBase encodes a single byte (8 bits) into a 128-bit RM(1,7) codeword.
// Returns two uint64 words [lo, hi] representing 128 bits.
func rmEncodeBase(msg byte) [2]uint64 {
	// The generator matrix rows for RM(1,7):
	// Row 0 (constant): all-ones = 0xFFFFFFFFFFFFFFFF (repeated)
	// Row 1: alternating 01010101... = 0xAAAAAAAAAAAAAAAA
	// Row 2: 00110011... = 0xCCCCCCCCCCCCCCCC
	// Row 3: 00001111... = 0xF0F0F0F0F0F0F0F0
	// Row 4: 0000000011111111... = 0xFF00FF00FF00FF00
	// Row 5: 16 zeros, 16 ones, ... = 0xFFFF0000FFFF0000
	// Row 6: 32 zeros, 32 ones, = 0xFFFFFFFF00000000
	//
	// For 128 bits we need two 64-bit words:
	// Row 7: first 64 zeros, next 64 ones = (0x0000000000000000, 0xFFFFFFFFFFFFFFFF)
	//
	// Message bit i selects row (i+1) in the traditional RM(1,m) generator matrix.
	// Bit 0 of msg is the constant term.

	var lo, hi uint64

	// bit0Mask expands bit 0 of msg to all 64 bits
	expand := func(bit uint) uint64 {
		return -(uint64((msg >> bit) & 1))
	}

	// Bit 0: constant row (all-ones if set)
	lo ^= expand(0)
	hi ^= expand(0)

	// Bit 1: row with pattern 0xAAAAAAAAAAAAAAAA
	lo ^= expand(1) & 0xAAAAAAAAAAAAAAAA
	hi ^= expand(1) & 0xAAAAAAAAAAAAAAAA

	// Bit 2: row with pattern 0xCCCCCCCCCCCCCCCC
	lo ^= expand(2) & 0xCCCCCCCCCCCCCCCC
	hi ^= expand(2) & 0xCCCCCCCCCCCCCCCC

	// Bit 3: row with pattern 0xF0F0F0F0F0F0F0F0
	lo ^= expand(3) & 0xF0F0F0F0F0F0F0F0
	hi ^= expand(3) & 0xF0F0F0F0F0F0F0F0

	// Bit 4: row with pattern 0xFF00FF00FF00FF00
	lo ^= expand(4) & 0xFF00FF00FF00FF00
	hi ^= expand(4) & 0xFF00FF00FF00FF00

	// Bit 5: row with pattern 0xFFFF0000FFFF0000
	lo ^= expand(5) & 0xFFFF0000FFFF0000
	hi ^= expand(5) & 0xFFFF0000FFFF0000

	// Bit 6: row with pattern 0xFFFFFFFF00000000
	lo ^= expand(6) & 0xFFFFFFFF00000000
	hi ^= expand(6) & 0xFFFFFFFF00000000

	// Bit 7: row (0x0000000000000000, 0xFFFFFFFFFFFFFFFF)
	hi ^= expand(7)

	return [2]uint64{lo, hi}
}

// rmDecode decodes an n2-bit received codeword (with duplicated RM(1,7))
// to a single byte using Walsh-Hadamard transform.
func rmDecode(src []uint64, n2, multiplicity int) byte {
	// Step 1: Accumulate all copies into a signed sum array of 128 entries.
	// Each bit position contributes +1 (if bit=1) or -1 (if bit=0).
	var sums [rmBaseLen]int32

	bitPos := 0
	for rep := 0; rep < multiplicity; rep++ {
		for i := 0; i < rmBaseLen; i++ {
			wordIdx := bitPos / 64
			bitIdx := uint(bitPos % 64)
			var bit int32
			if wordIdx < len(src) {
				bit = int32((src[wordIdx] >> bitIdx) & 1)
			}
			// Convert 0/1 to +1/-1 (0 -> +1, 1 -> -1)
			sums[i] += 1 - 2*bit
			bitPos++
		}
	}

	// Step 2: Fast Walsh-Hadamard Transform (in-place, 7 passes)
	for pass := 0; pass < 7; pass++ {
		step := 1 << uint(pass)
		for i := 0; i < rmBaseLen; i += 2 * step {
			for j := i; j < i+step; j++ {
				a := sums[j]
				b := sums[j+step]
				sums[j] = a + b
				sums[j+step] = a - b
			}
		}
	}

	// Step 3: Find the position with maximum absolute value
	maxAbs := int32(0)
	maxPos := 0
	sign := int32(1)

	for i := 0; i < rmBaseLen; i++ {
		v := sums[i]
		abs := v
		if abs < 0 {
			abs = -abs
		}
		if abs > maxAbs {
			maxAbs = abs
			maxPos = i
			if v > 0 {
				sign = 1
			} else {
				sign = -1
			}
		}
	}

	// Step 4: Recover the message byte.
	// maxPos encodes bits 1-7 of the message, sign encodes bit 0.
	// The codeword at position j is: c[j] = m0 XOR m1*j[0] XOR m2*j[1] ... XOR m7*j[6]
	// The WHT peak at position p means (m1,..,m7) = bits of p.
	// Negative sign means m0 = 1 (constant all-ones was applied).
	msg := byte(maxPos << 1)
	if sign < 0 {
		msg |= 1
	}
	return msg
}
